internal/extsvc/npm/npm.go

```
@@ -0,0 +1,242 @@
+}
+
+func (i illFormedJSONError) Error() string {
+	return fmt.Sprintf("unexpected JSON output from NPM request: url=%s", i.url)
+}
+
+func (client *HTTPClient) do(ctx context.Context, req *http.Request) (*http.Response, error) {
+	req, ht := nethttp.TraceRequest(ot.GetTracer(ctx),
+		req.WithContext(ctx),
+		nethttp.OperationName("NPM"),
+		nethttp.ClientTrace(false))
+	defer ht.Finish()
+	startWait := time.Now()
+	if err := client.limiter.Wait(ctx); err != nil {
+		return nil, err
+	}
+	if d := time.Since(startWait); d > 200*time.Millisecond {
+		log15.Warn("NPM self-enforced API rate limit: request delayed longer than expected due to rate limit", "delay", d)
+	}
+	return client.doer.Do(req)
+}
+
+func (client *HTTPClient) makeGetRequest(ctx context.Context, url string) (io.ReadSeeker, error) {
+	var bodyBuffer bytes.Reader
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, &bodyBuffer)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	resp.Body.Close()
+	return &bodyBuffer, nil
+}
+
+func (client *HTTPClient) getDependencyInfo(ctx context.Context, dep reposource.NPMDependency) (info npmDependencyInfo, err error) {
+	// https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#getpackageversion
+	url := fmt.Sprintf("%s/%s/%s", client.registryURL, dep.Package.PackageSyntax(), dep.Version)
+	reader, err := client.makeGetRequest(ctx, url)
+	if err != nil {
+		return info, err
+	}
+	body, err := io.ReadAll(reader)
+	if err != nil {
+		return info, errors.Wrapf(err, "failed to read NPM's response: url=%s", url)
+	}
+	if json.Unmarshal(body, &info) != nil {
+		return info, illFormedJSONError{url: url}
+	}
+	return info, nil
+}
+
+func (client *HTTPClient) DoesDependencyExist(ctx context.Context, dep reposource.NPMDependency) (exists bool, err error) {
+	_, err = client.getDependencyInfo(ctx, dep)
+	var e illFormedJSONError
+	if errors.As(err, &e) {
+		return false, nil
+	}
```

Do we want to silently fail here or should we log `e`?

---

Added a `log15.Warn` call for the error.